verification/spec/connection-handshake/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,353 @@
+    /\ VerifyConnProof(m.connProof)
+
+
+(* Handles a "ICS3MsgConfirm" message.
+ *)
+HandleConfirmMsg(m) ==
+    LET newCon == [parameters |-> m.parameters,
+                   state |-> "OPEN"] IN 
+    IF PreconditionsConfirmMsg(m)
+    THEN [out |-> outBuf,    (* Never need to reply to a confirm msg. *)
+          store |-> NewStore(newCon)]
+    ELSE [out |-> outBuf,
+          store |-> store]
+
+
+(* Action for advancing the current height (latestHeight) of the chain.
+
+    The environment triggers this as part of the GoodNextEnv action.
+ *) 
+AdvanceChainHeight ==
+    store' = [store EXCEPT !.latestHeight = @ + 1]
+
+
+(* State predicate returning true if MaxChainHeight not yet attained.
+ *)
+CanAdvance ==
+    store.latestHeight < MaxChainHeight
+
+
+(* Action for updating the local client on this chain with a height.
+    
+    The environment triggers this as part of the GoodNextEnv action.
+    This primes the store; leaves the chain buffers unchanged.
+    This will also advance the chain height.
+ *)
+UpdateClient(height) ==
+    \/ /\ height \notin store.client.consensusStates
+       /\ store' = [store EXCEPT !.latestHeight = @ + 1,
+                                 !.client.consensusStates = @ \cup {height},
+                                 !.client.latestHeight = height]
+    \/ /\ height \in store.client.consensusStates
+       /\ UNCHANGED store
+
+
+(* Generic action for handling any type of inbound message.
+
+    Expects as parameter a message.
+    Takes care of invoking priming the 'store' and any reply msg in 'outBuf'.
+    This action assumes the message type is valid, therefore one of the
+    disjunctions will always enable.
+ *)
+ProcessMsg(m) ==
+    LET res == CASE m.type = "ICS3MsgInit" -> HandleInitMsg(m)
+                 [] m.type = "ICS3MsgTry" -> HandleTryMsg(m)
+                 [] m.type = "ICS3MsgAck" -> HandleAckMsg(m)
+                 [] m.type = "ICS3MsgConfirm" -> HandleConfirmMsg(m) IN
+    /\ outBuf' = res.out
+    /\ store' = res.store
+
+
+(***************************************************************************
+ Connection Handshake Module (ICS3) main spec.
+ ***************************************************************************)
+
+
+Init(chainID, client, connection) ==
+    /\ store = [id |-> chainID,
+                latestHeight |-> 1,
+                connection |-> connection,
+                client |-> client]
+
+
+Next ==
+    \/ /\ inBuf # <<>>            \* Enabled if we have an inbound msg.
+       /\ CanAdvance
+       /\ ProcessMsg(Head(inBuf)) \* Generic action for handling a msg.
+       /\ inBuf' = Tail(inBuf)    \* Strip the head of our inbound msg. buffer.
+    \/ /\ inBuf = <<>>
+       /\ ~ CanAdvance
+       /\ UNCHANGED<<moduleVars>>
+
+
+TypeInvariant ==
+    /\ inBuf \in Seq(ConnectionHandshakeMessages) \union {<<>>}
+    /\ outBuf \in Seq(ConnectionHandshakeMessages) \union {<<>>}
+    /\ store.connection \in Connections
```

Comment:
You can define for a store what type it should have: see for example how it was done in fastsync for complex record https://github.com/informalsystems/tendermint-rs/blob/02efca7eb50fc2ca6233038a7bd97b21ad6ab577/docs/spec/fastsync/fastsync.tla#L634.

Reply:
Thanks for the pointer, added that.